<p align="center">
<img src='../../img/VerticaMLPython.png' width="180px">
</p>

# Vertica ML Python Exercise 4

During this exercice, we will:
<ul>
    <li> Normalize the data using the different methods
    <li> Create a PCA model
    <li> Find a suitable number of principal components to keep 
</ul>
## Initialization

Let's create a cursor using the vertica_cursor function

In [2]:
from vertica_ml_python.utilities import vertica_cursor
cur = vertica_cursor("VerticaDSN")

During this study, we will work with the 'heart diseases' dataset introduced in exercise 2. 

In [3]:
from vertica_ml_python import vDataframe
heart = vDataframe('heart', cursor = cur)
print(heart)

<object>  Name: heart, Number of rows: 270, Number of columns: 14


This dataset contains many information of 270 patients including:
<ul>
    <li><b>age:</b> age in years</li>
    <li><b>thalach:</b> maximum heart rate achieved</li>
    <li><b>trestbps:</b> resting blood pressure (in mm Hg on admission to the hospital) </li>
    <li><b>fbs:</b> (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false) </li>
    <li><b>slope:</b> the slope of the peak exercise ST segment (Value 1: upsloping; Value 2: flat; Value 3: downsloping) </li>
    <li><b>ca:</b> number of major vessels (0-3) colored by flourosopy </li>
    <li><b>num:</b> diagnosis of heart disease (angiographic disease status) (Value 1: < 50% diameter narrowing; Value 2: > 50% diameter narrowing)</li>
    <li><b>sex:</b> sex (1 = male; 0 = female) </li>
    <li><b>cp:</b> hest pain type (Value 1: typical angina; Value 2: atypical angina; Value 3: non-anginal pain; Value 4: asymptomatic) </li>
    <li><b>exang:</b> exercise induced angina (1 = yes; 0 = no) </li>
    <li><b>restecg:</b> resting electrocardiographic results (Value 0: normal; Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV); Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria) </li>
    <li><b>chol:</b> serum cholestoral in mg/dl </li>
    <li><b>oldpeak:</b> ST depression induced by exercise relative to rest </li>
    <li><b>thal:</b> 3 = normal; 6 = fixed defect; 7 = reversable defect </li>
</ul>

The purpose is to find people having heart complications compare to the rest. 

## Decomposition and Normalization

Let's explore the data by displaying descriptive statistics of all the columns.

In [4]:
heart.describe(method = "categorical")

<object>

In [5]:
heart.describe(method = "numerical")

<object>

Many features are categorical and need to be encoded first. However, it is not needed during this exercice as the main purpose is to understand normalization and decomposition. We will only use numerical features during this study (age, thalach, trestbps, chol and oldpeak).

In exercise 2, we didn't normalize the data as we chose features which were approximately living in the same interval. However, it is important to normalize the data when we are dealing with algorithms using a notion of distance during the process. 

<b>Question 1: </b>Look at the descriptive statistics of all the elements and compare the median and the average. What do you notice ? Explain why the 'zscore' is a suitable way to normalize the data. Normalize the data.

<b>Question 2: </b>If we decide to build a model, these columns could be correlated one to another. PCA allows a space reduction and create independent components. Create a PCA model using these features.

<b>Question 3: </b>Look at the explained_variance attribute and choose a number of components to keep (we usually consider to keep components which bring the accumulated explained variance to more than 0.95)

<b>Question 4: </b>Why is it important to reduce the number of features before fitting a ML model ?